In [113]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder

In [114]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

In [115]:
#review train_df
train_df.head()

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0       57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1      141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2      321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3       11778   11778     3000.0    0.1240       100.22           RENT   
4      169382  169382    30000.0    0.1612      1056.49       MORTGAGE   

   annual_inc verification_status loan_status pymnt_plan  ...  pct_tl_nvr_dlq  \
0    223000.0        Not Verified    low_risk          n  ...           100.0   
1    123000.0     Source Verified    low_risk          n  ...            85.0   
2    197000.0     Source Verified    low_risk          n  ...            85.7   
3     45000.0        Not Verified    low_risk          n  ...           100.0   
4    133000.0     Source Verified    low_risk          n  ...           100.0   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              50.0                   0.0        0.0         577150.0   
1              33.3                   0.0        0.0         132750.0   
2              33.3                   0.0        0.0         628160.0   
3              16.7                   1.0        0.0          42006.0   
4              66.7                   0.0        0.0         283248.0   

   total_bal_ex_mort  total_bc_limit total_il_high_credit_limit  \
0           122018.0         32000.0                   170200.0   
1            27896.0         15900.0                    35398.0   
2           114043.0         22600.0                    90340.0   
3            20761.0         19900.0                    15406.0   
4           109056.0         79500.0                    58778.0   

   hardship_flag  debt_settlement_flag  
0              N                     N  
1              N                     N  
2              N                     N  
3              N                     N  
4              N                     N  

[5 rows x 86 columns]

In [116]:
#review test_df
test_df.head()

Unnamed: 0  index  loan_amnt  int_rate  installment home_ownership  \
0       67991  67991    40000.0    0.0819       814.70       MORTGAGE   
1       25429  25429     6000.0    0.1524       208.70           RENT   
2       38496  38496     3600.0    0.1695       128.27           RENT   
3       19667  19667    20000.0    0.1524       478.33           RENT   
4       37505  37505     3600.0    0.1240       120.27           RENT   

   annual_inc verification_status loan_status pymnt_plan  ...  pct_tl_nvr_dlq  \
0    140000.0        Not Verified    low_risk          n  ...            97.7   
1     55000.0        Not Verified    low_risk          n  ...            66.7   
2     42000.0        Not Verified    low_risk          n  ...           100.0   
3    100000.0        Not Verified    low_risk          n  ...           100.0   
4     50000.0        Not Verified    low_risk          n  ...           100.0   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0               0.0                   0.0        0.0         527975.0   
1               0.0                   0.0        0.0          34628.0   
2               0.0                   0.0        0.0          23100.0   
3              50.0                   0.0        0.0          56481.0   
4              25.0                   0.0        0.0          45977.0   

   total_bal_ex_mort  total_bc_limit total_il_high_credit_limit  \
0            70914.0         74600.0                    99475.0   
1            23460.0          5900.0                    23628.0   
2            19183.0          7300.0                    15000.0   
3            43817.0         13800.0                    35981.0   
4            32448.0         21000.0                    24977.0   

   hardship_flag  debt_settlement_flag  
0              N                     N  
1              N                     N  
2              N                     N  
3              N                     N  
4              N                     N  

[5 rows x 86 columns]

In [117]:
# Convert categorical data to numeric and separate target feature for training data
#convert data using get.dummies
#Drop the taget column and any unneeded columns
X = train_df.drop(['Unnamed: 0', 'loan_status'], axis=1)
X

index  loan_amnt  int_rate  installment home_ownership  annual_inc  \
0       57107    13375.0    0.1797       483.34       MORTGAGE    223000.0   
1      141451    21000.0    0.1308       478.68       MORTGAGE    123000.0   
2      321143    20000.0    0.1240       448.95       MORTGAGE    197000.0   
3       11778     3000.0    0.1240       100.22           RENT     45000.0   
4      169382    30000.0    0.1612      1056.49       MORTGAGE    133000.0   
...       ...        ...       ...          ...            ...         ...   
12175  354912    19975.0    0.2565       801.09           RENT     28000.0   
12176  354944    15000.0    0.1774       540.34           RENT     50000.0   
12177  354973     3600.0    0.1862       131.28           RENT     60000.0   
12178  355002    15000.0    0.0881       475.68       MORTGAGE     62000.0   
12179  355307    15000.0    0.1774       540.34           RENT     54080.0   

      verification_status pymnt_plan    dti  delinq_2yrs  ...  pct_tl_nvr_dlq  \
0            Not Verified          n  29.99          0.0  ...           100.0   
1         Source Verified          n  11.26          2.0  ...            85.0   
2         Source Verified          n  11.28          0.0  ...            85.7   
3            Not Verified          n  18.08          0.0  ...           100.0   
4         Source Verified          n  27.77          0.0  ...           100.0   
...                   ...        ...    ...          ...  ...             ...   
12175        Not Verified          n  28.42          0.0  ...           100.0   
12176            Verified          n  23.43          4.0  ...            90.5   
12177        Not Verified          n  28.80          0.0  ...           100.0   
12178     Source Verified          n  11.44          0.0  ...           100.0   
12179        Not Verified          n  21.66          1.0  ...            90.0   

       percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0                  50.0                   0.0        0.0         577150.0   
1                  33.3                   0.0        0.0         132750.0   
2                  33.3                   0.0        0.0         628160.0   
3                  16.7                   1.0        0.0          42006.0   
4                  66.7                   0.0        0.0         283248.0   
...                 ...                   ...        ...              ...   
12175              16.7                   0.0        0.0          50055.0   
12176              11.1                   0.0        0.0          70324.0   
12177               0.0                   0.0        0.0          83765.0   
12178               0.0                   0.0        0.0         189930.0   
12179             100.0                   0.0        0.0          48968.0   

      total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0              122018.0         32000.0                    170200.0   
1               27896.0         15900.0                     35398.0   
2              114043.0         22600.0                     90340.0   
3               20761.0         19900.0                     15406.0   
4              109056.0         79500.0                     58778.0   
...                 ...             ...                         ...   
12175           28192.0         18700.0                     19055.0   
12176           57025.0         13300.0                     54824.0   
12177           55156.0         14800.0                     53065.0   
12178           23748.0          7000.0                     32930.0   
12179           29786.0          2900.0                     33968.0   

       hardship_flag  debt_settlement_flag  
0                  N                     N  
1                  N                     N  
2                  N                     N  
3                  N                     N  
4                  N                     N  
...              ...                   ...  
12175

In [118]:
#convert data using get.dummies
X_train = pd.get_dummies(X)
print(X_train.columns)
X_train.head()

Index(['index', 'loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_

index  loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0   57107    13375.0    0.1797       483.34    223000.0  29.99          0.0   
1  141451    21000.0    0.1308       478.68    123000.0  11.26          2.0   
2  321143    20000.0    0.1240       448.95    197000.0  11.28          0.0   
3   11778     3000.0    0.1240       100.22     45000.0  18.08          0.0   
4  169382    30000.0    0.1612      1056.49    133000.0  27.77          0.0   

   inq_last_6mths  open_acc  pub_rec  ...  verification_status_Verified  \
0             0.0      15.0      0.0  ...                             0   
1             0.0      16.0      0.0  ...                             0   
2             0.0      12.0      0.0  ...                             0   
3             0.0      12.0      1.0  ...                             0   
4             2.0      13.0      0.0  ...                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                1   
1                            1                           0                1   
2                            1                           0                1   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  debt_settlement_flag_Y  
0                0                       1                       0  
1                0                       1                       0  
2                0                       1                       0  
3                0                       1                       0  
4                0                       1                       0  

[5 rows x 93 columns]

In [119]:
# Converting output labels to 0 and 1 using LabelEncoder
y_train = LabelEncoder().fit_transform(train_df['loan_status'])
y_train

array([1, 1, 1, ..., 0, 0, 0])

In [120]:
# Convert categorical data to numeric and separate target feature for testing data
X = test_df.drop(['Unnamed: 0', 'loan_status'], axis=1)

X

index  loan_amnt  int_rate  installment home_ownership  annual_inc  \
0     67991    40000.0    0.0819       814.70       MORTGAGE    140000.0   
1     25429     6000.0    0.1524       208.70           RENT     55000.0   
2     38496     3600.0    0.1695       128.27           RENT     42000.0   
3     19667    20000.0    0.1524       478.33           RENT    100000.0   
4     37505     3600.0    0.1240       120.27           RENT     50000.0   
...     ...        ...       ...          ...            ...         ...   
4697  77282    30000.0    0.1240       673.42           RENT    140480.0   
4698  77291    24000.0    0.0756       747.22           RENT     50000.0   
4699  77292    10000.0    0.2305       387.36           RENT     33000.0   
4700  77297     8000.0    0.1862       205.86           RENT     38000.0   
4701  77304    30000.0    0.2055      1123.34           RENT    180000.0   

     verification_status pymnt_plan    dti  delinq_2yrs  ...  pct_tl_nvr_dlq  \
0           Not Verified          n  19.75          0.0  ...            97.7   
1           Not Verified          n  11.52          2.0  ...            66.7   
2           Not Verified          n   6.74          0.0  ...           100.0   
3           Not Verified          n  12.13          0.0  ...           100.0   
4           Not Verified          n  16.08          0.0  ...           100.0   
...                  ...        ...    ...          ...  ...             ...   
4697     Source Verified          n  15.74          0.0  ...           100.0   
4698        Not Verified          n  26.81          0.0  ...           100.0   
4699            Verified          n  38.51          0.0  ...           100.0   
4700     Source Verified          n  16.36          0.0  ...            95.0   
4701     Source Verified          n  12.06          0.0  ...           100.0   

      percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0                  0.0                   0.0        0.0         527975.0   
1                  0.0                   0.0        0.0          34628.0   
2                  0.0                   0.0        0.0          23100.0   
3                 50.0                   0.0        0.0          56481.0   
4                 25.0                   0.0        0.0          45977.0   
...                ...                   ...        ...              ...   
4697              28.6                   0.0        0.0         159688.0   
4698               0.0                   0.0        0.0          62375.0   
4699               0.0                   0.0        0.0          43250.0   
4700               0.0                   1.0        0.0          31357.0   
4701              33.3                   0.0        0.0         218686.0   

     total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0              70914.0         74600.0                     99475.0   
1              23460.0          5900.0                     23628.0   
2              19183.0          7300.0                     15000.0   
3              43817.0         13800.0                     35981.0   
4              32448.0         21000.0                     24977.0   
...                ...             ...                         ...   
4697          110873.0         48400.0                    107388.0   
4698           18928.0         13300.0                     30775.0   
4699           33022.0          8500.0                     29550.0   
4700           19595.0          1500.0                      9657.0   
4701          209389.0          5600.0                    209986.0   

      hardship_flag  debt_settlement_flag  
0                 N                     N  
1                 N                     N  
2                 N                     N  
3                 N                     N  
4                 N                     N  
...             ...                   ...  
4697              N                     N  
4698              N       

In [121]:
#convert data using get.dummies
X_test = pd.get_dummies(X)
print(X_test.columns)
X_test.head()

Index(['index', 'loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_

index  loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0  67991    40000.0    0.0819       814.70    140000.0  19.75          0.0   
1  25429     6000.0    0.1524       208.70     55000.0  11.52          2.0   
2  38496     3600.0    0.1695       128.27     42000.0   6.74          0.0   
3  19667    20000.0    0.1524       478.33    100000.0  12.13          0.0   
4  37505     3600.0    0.1240       120.27     50000.0  16.08          0.0   

   inq_last_6mths  open_acc  pub_rec  ...  \
0             1.0      18.0      0.0  ...   
1             0.0       8.0      0.0  ...   
2             0.0       6.0      0.0  ...   
3             2.0       7.0      0.0  ...   
4             3.0       6.0      0.0  ...   

   verification_status_Source Verified  verification_status_Verified  \
0                                    0                             0   
1                                    0                             0   
2                                    0                             0   
3                                    0                             0   
4                                    0                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                1   
1                            1                           0                1   
2                            1                           0                1   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  
0                0                       1  
1                0                       1  
2                0                       1  
3                0                       1  
4                0                       1  

[5 rows x 92 columns]

In [122]:
#Find missing test column
set(X_train.columns) - set(X_test.columns)

{'debt_settlement_flag_Y'}

In [101]:
# add missing dummy variables to testing set
X_test["debt_settlement_flag_Y"] = 0

X_test.head()

index  loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0  67991    40000.0    0.0819       814.70    140000.0  19.75          0.0   
1  25429     6000.0    0.1524       208.70     55000.0  11.52          2.0   
2  38496     3600.0    0.1695       128.27     42000.0   6.74          0.0   
3  19667    20000.0    0.1524       478.33    100000.0  12.13          0.0   
4  37505     3600.0    0.1240       120.27     50000.0  16.08          0.0   

   inq_last_6mths  open_acc  pub_rec  ...  verification_status_Verified  \
0             1.0      18.0      0.0  ...                             0   
1             0.0       8.0      0.0  ...                             0   
2             0.0       6.0      0.0  ...                             0   
3             2.0       7.0      0.0  ...                             0   
4             3.0       6.0      0.0  ...                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                1   
1                            1                           0                1   
2                            1                           0                1   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  debt_settlement_flag_Y  
0                0                       1                       0  
1                0                       1                       0  
2                0                       1                       0  
3                0                       1                       0  
4                0                       1                       0  

[5 rows x 93 columns]

In [102]:
# Converting output labels to 0 and 1 using LabelEncoder
y_test = LabelEncoder().fit_transform(test_df['loan_status'])
y_test


array([1, 1, 1, ..., 0, 0, 0])

# Model Prediction
My prediction is that the Random Forest Classifier model will perform better. I believe this will be true for the following reasons:
 - There are a number of categorical columns in the dataset and a Random Forest Classifier may perform better
 - There are a number of variables in the dataset and a Random Forest Classifier may perform better and capture feature patterns
 - With Random Forest Classifier you could also feature selection to select only the relevant values in the dataset 

In [103]:
# Train the Logistic Regression model on the unscaled data and print the model score
from sklearn.linear_model import LogisticRegression

classifier.fit(X_train,y_train)
print("Testing Score:", classifier.score(X_test, y_test))
print("Training Score:", classifier.score(X_train, y_train))

Testing Score: 0.5691195236069757
Training Score: 0.7071428571428572


In [104]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(random_state=13, n_estimators=300).fit(X_train, y_train)
print(f'Testing Score: {clf.score(X_test, y_test)}')
print(f'Training Score: {clf.score(X_train, y_train)}')

Testing Score: 0.6897065078689919
Training Score: 1.0


# Impact of Scaling
My assumption is that scaling the data will improve the accuracy and performance of both models for the following reasons:
- Reducing the magnitude of outliers or large values that can impact the accuracy of the model due the large size
- There is a large range of values in the dataset. The machine learning algorithm does not know what the number represents and could make erroneous assumptions if not scaled.


In [107]:
# Scale the data
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [123]:
# Train the Logistic Regression model on the scaled data and print the model score
classifiers = LogisticRegression(solver="liblinear")

classifiers.fit(X_train_scaled, y_train)
print("Testing Score:", classifiers.score(X_test_scaled, y_test))
print("Training Score:", classifiers.score(X_train_scaled, y_train))

Testing Score: 0.719906422798809
Training Score: 0.7126436781609196


In [124]:
# Train a Random Forest Classifier model on the scaled data and print the model score
clfs = RandomForestClassifier(random_state=1, n_estimators=300).fit(X_train_scaled, y_train)
print(f'Testing Score: {clfs.score(X_test_scaled, y_test)}')
print(f'Training Score: {clfs.score(X_train_scaled, y_train)}')

Testing Score: 0.6573798383666525
Training Score: 1.0


# Model Comparisons

## Performance
Prior to scaling, the models perform similarly with an edge to Random Forest. With Random Forest the accuracy is at 65% for the testing score and is 100% for the training score, which raises concerns of over-fitting.

## Scaling Improvements
Surprisingly, scaling the data had an impact on the Testing Score of the Logistic Regression model but it did not seem to have much of an impact on the Random Forest Classifier. Based on Google research, Random Forest does not necessarily need scaling since the nature of the model does not present the numerical precision issues found with Logistic Regression.

## Final Thoughts
I am going to change my position on the models and give the Logistic Regression a slight edge over Random Forest. After scaling the model performs better than Random Forest with a testing score of 72 % over the Random Forest testing score of 66%. Random Forest has a perfect training score, 100%, but I have concerns with over-fitting. Perhaps if feature selection was applied the chances of over-fitting could be reduced.
